File: claudsys.py
import http.server
import os
import socket
import subprocess
import sys
import threading
from datetime import datetime

APP_SCRIPT = os.path.join(os.path.abspath(os.path.dirname(__file__)), "main.py")
DEFAULT_PORT = int(os.environ.get("CLAUDSYS_PORT", "8000"))


class LauncherState:
    def __init__(self):
        self.process = None
        self.lock = threading.Lock()

    def status(self):
        with self.lock:
            if self.process is None:
                return "idle"
            if self.process.poll() is None:
                return "running"
            self.process = None
            return "idle"

    def launch(self):
        with self.lock:
            if self.process is None or self.process.poll() is not None:
                self.process = subprocess.Popen([sys.executable, APP_SCRIPT])
                return True
            return False


state = LauncherState()


class ClaudsysHandler(http.server.BaseHTTPRequestHandler):
    def _html(self, body: str) -> bytes:
        page = f"""
        <!DOCTYPE html>
        <html lang=\"tr\">
        <head>
            <meta charset=\"UTF-8\" />
            <title>Claudsys Launcher</title>
            <style>
                body {{ font-family: Arial, sans-serif; background:#f4f4f7; color:#1f2933; margin:0; padding:0; }}
                header {{ background:#1f2933; color:#fefefe; padding:16px 24px; display:flex; justify-content:space-between; align-items:center; }}
                main {{ max-width:720px; margin:40px auto; padding:24px; background:#fff; border-radius:12px; box-shadow:0 4px 18px rgba(0,0,0,0.08); }}
                button {{ background:#2563eb; border:none; color:#fff; padding:12px 20px; font-size:16px; border-radius:8px; cursor:pointer; }}
                button:disabled {{ background:#94a3b8; cursor:not-allowed; }}
                .status {{ margin-top:16px; font-weight:600; }}
                footer {{ text-align:center; margin-top:32px; color:#52606d; font-size:14px; }}
            </style>
        </head>
        <body>
            <header>
                <div><strong>Claudsys</strong> — LAN Launcher</div>
                <div>{datetime.now().strftime('%d-%m-%Y %H:%M:%S')}</div>
            </header>
            <main>
                <h1>STS-SpindleTakipSistemi</h1>
                <p>Launch the desktop GUI on this host so others on your LAN can use it.</p>
                {body}
                <footer>Serving on {self.server.server_address[0]}:{self.server.server_address[1]}</footer>
            </main>
        </body>
        </html>
        """
        return page.encode("utf-8")

    def do_GET(self):
        status = state.status()
        button_state = "disabled" if status == "running" else ""
        message = "GUI is already running." if status == "running" else "Click launch to start the GUI." 
        body = f"""
            <form method=\"post\" action=\"/launch\">
                <button type=\"submit\" {button_state}>Launch STS GUI</button>
            </form>
            <div class=\"status\">Status: {status}</div>
            <p>{message}</p>
        """
        content = self._html(body)
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(content)))
        self.end_headers()
        self.wfile.write(content)

    def do_POST(self):
        if self.path != "/launch":
            self.send_error(404)
            return
        launched = state.launch()
        if launched:
            body = "<div class=\"status\">Launched STS GUI.</div>"
        else:
            body = "<div class=\"status\">GUI already running.</div>"
        content = self._html(body)
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(content)))
        self.end_headers()
        self.wfile.write(content)

    def log_message(self, format, *args):  # noqa: A003
        return


def run_server(port=DEFAULT_PORT):
    server_address = ("0.0.0.0", port)
    httpd = http.server.ThreadingHTTPServer(server_address, ClaudsysHandler)
    host_ip = socket.gethostbyname(socket.gethostname())
    print(f"Claudsys running on http://{host_ip}:{port} (LAN) and http://localhost:{port}")
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        httpd.server_close()


if __name__ == "__main__":
    run_server()

File: claudsys.js
const http = require('http');
const os = require('os');
const path = require('path');
const { spawn } = require('child_process');

const APP_SCRIPT = path.join(__dirname, 'main.py');
const DEFAULT_PORT = parseInt(process.env.CLAUDSYS_PORT || '8000', 10);
const PYTHON_CMD = process.env.CLAUDSYS_PYTHON || process.env.PYTHON || 'python';

let childProcess = null;

function status() {
  if (childProcess && childProcess.exitCode === null) {
    return 'running';
  }
  childProcess = null;
  return 'idle';
}

function launchGui() {
  if (status() === 'running') {
    return false;
  }
  try {
    childProcess = spawn(PYTHON_CMD, [APP_SCRIPT], { detached: true, stdio: 'ignore' });
    childProcess.unref();
    childProcess.on('exit', () => {
      childProcess = null;
    });
    childProcess.on('error', () => {
      childProcess = null;
    });
    return true;
  } catch (err) {
    console.error('Failed to launch GUI:', err);
    childProcess = null;
    return false;
  }
}

function render(body, serverAddress) {
  const now = new Date();
  const timestamp = now.toLocaleString('tr-TR', {
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  });

  return `<!DOCTYPE html>
  <html lang="tr">
  <head>
    <meta charset="UTF-8" />
    <title>Claudsys Launcher</title>
    <style>
      body { font-family: Arial, sans-serif; background:#f4f4f7; color:#1f2933; margin:0; padding:0; }
      header { background:#1f2933; color:#fefefe; padding:16px 24px; display:flex; justify-content:space-between; align-items:center; }
      main { max-width:720px; margin:40px auto; padding:24px; background:#fff; border-radius:12px; box-shadow:0 4px 18px rgba(0,0,0,0.08); }
      button { background:#2563eb; border:none; color:#fff; padding:12px 20px; font-size:16px; border-radius:8px; cursor:pointer; }
      button:disabled { background:#94a3b8; cursor:not-allowed; }
      .status { margin-top:16px; font-weight:600; }
      footer { text-align:center; margin-top:32px; color:#52606d; font-size:14px; }
    </style>
  </head>
  <body>
    <header>
      <div><strong>Claudsys</strong> — LAN Launcher</div>
      <div>${timestamp}</div>
    </header>
    <main>
      <h1>STS-SpindleTakipSistemi</h1>
      <p>Bu makinede masaüstü GUI'yi başlatmak için tarayıcıdan tıklayın.</p>
      ${body}
      <footer>Serving on ${serverAddress}</footer>
    </main>
  </body>
  </html>`;
}

function getLanAddress(port) {
  const nets = os.networkInterfaces();
  for (const name of Object.keys(nets)) {
    for (const net of nets[name] || []) {
      if (net.family === 'IPv4' && !net.internal) {
        return `${net.address}:${port}`;
      }
    }
  }
  return `localhost:${port}`;
}

function handleRequest(req, res, port) {
  if (req.method === 'GET' && req.url === '/') {
    const currentStatus = status();
    const disabled = currentStatus === 'running' ? 'disabled' : '';
    const message = currentStatus === 'running' ? 'GUI zaten çalışıyor.' : 'GUI\'yi başlatmak için tıklayın.';
    const body = `
      <form method="post" action="/launch">
        <button type="submit" ${disabled}>STS GUI\'yi Başlat</button>
      </form>
      <div class="status">Durum: ${currentStatus}</div>
      <p>${message}</p>
    `;
    const html = render(body, getLanAddress(port));
    const data = Buffer.from(html, 'utf-8');
    res.writeHead(200, {
      'Content-Type': 'text/html; charset=utf-8',
      'Content-Length': data.length,
    });
    res.end(data);
    return;
  }

  if (req.method === 'POST' && req.url === '/launch') {
    let launched = false;
    try {
      launched = launchGui();
    } catch (err) {
      console.error('Launch error:', err);
    }
    const body = `<div class="status">${launched ? 'STS GUI başlatıldı.' : 'GUI zaten çalışıyor veya başlatılamadı.'}</div>`;
    const html = render(body, getLanAddress(port));
    const data = Buffer.from(html, 'utf-8');
    res.writeHead(200, {
      'Content-Type': 'text/html; charset=utf-8',
      'Content-Length': data.length,
    });
    res.end(data);
    return;
  }

  res.writeHead(404, { 'Content-Type': 'text/plain; charset=utf-8' });
  res.end('Not Found');
}

function startServer(port = DEFAULT_PORT) {
  const server = http.createServer((req, res) => handleRequest(req, res, port));
  server.listen(port, '0.0.0.0', () => {
    console.log(`Claudsys running on http://${getLanAddress(port)} (LAN) and http://localhost:${port}`);
  });

  server.on('error', (err) => {
    console.error('Server error:', err);
  });
}

if (require.main === module) {
  startServer();
}

module.exports = { startServer };
